### PR TITLE
Feature/user registration confirmation

### PR DESCRIPTION
**Pull request by DandyCV**

### CODE REVIEW

File: app/controllers/concerns/response.rb [hunk 1]
```diff
@@ -1,7 +1,11 @@
 # frozen_string_literal: true
 
 module Response
-  def respond_with(entity:, status: 200, serializer: Api::V1::Serializer::Error)
-    render json: serializer.new(entity).to_json, status: status
+  def respond_with(entity: nil, status: 200, serializer: Api::V1::Serializer::Error)
+    if entity
+      render json: serializer.new(entity).to_json, status: status
+    else
+      render json: {}, status: status
```
Comment (bestwebua) [inline app/controllers/concerns/response.rb]:
```ruby
head(:ok)
```

File: spec/requests/v1/users/confirmations_spec.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  let(:email_token) { Api::V1::Lib::Service::EmailToken.encode(user_data) }
+  let(:params) { { email_token: email_token } }
+
+  describe 'GET /api/v1/users/confirmation' do
+    before { get "/api/v1/users/confirmation?email_token=#{email_token}", as: :json }
+
+    describe 'Succes' do
+      it 'renders OK' do
+        expect(response).to be_ok
+        expect(JSON.parse(response.body)).to be_empty
```
Comment (bestwebua) [inline spec/requests/v1/users/confirmations_spec.rb]:
```ruby
expect(response).to be_no_content
```

File: spec/requests/v1/users/confirmations_spec.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        expect(User.find(user.id)).to be_confirmed
```
Comment (bestwebua) [inline spec/requests/v1/users/confirmations_spec.rb]: Request test is about http statuses and rendered context expectations. What are you doing here?

**Review by bestwebua — CHANGES_REQUESTED** [review]

File: app/controllers/concerns/response.rb (hunk 1 above)
Comment (DandyCV) [reply app/controllers/concerns/response.rb]: Ok

File: spec/requests/v1/users/confirmations_spec.rb (hunk 2 above)
Comment (DandyCV) [reply spec/requests/v1/users/confirmations_spec.rb]: In previous comment we render :ok but now we expect :no_content, so what to chose for updating `confirmed` field in  User record?

File: spec/requests/v1/users/confirmations_spec.rb (hunk 3 above)
Comment (DandyCV) [reply spec/requests/v1/users/confirmations_spec.rb]: I agree, my bad!

File: app/concepts/api/v1/users/confirmations/operation/confirm.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+module Api::V1::Users::Confirmations::Operation
+  class Confirm < ApplicationOperation
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/operation/confirm.rb]: Show

File: app/concepts/api/v1/users/confirmations/operation/confirm.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    def decode_token(params)
+      data = Api::V1::Lib::Service::EmailToken.decode(params[:email_token])
+    rescue JWT::DecodeError
+      Failure({ errors: [{ token: I18n.t('users.confirmations.token.invalid') }] })
+    else
+      Success(data.first)
+    end
+
+    def validate_token(data)
+      if (DateTime.now - Constants::Shared::VALID_TOKEN_TIME) < DateTime.parse(data['created_at'])
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/operation/confirm.rb]: Please reverse condition.

File: spec/concepts/api/v1/serializer/error_spec.rb [hunk 6]
```diff
@@ -8,23 +8,34 @@
...
+      it 'uses memoized errors' do
+        expect_any_instance_of(described_class).to receive(:errors).once.and_call_original
+        2.times { error_serializer }
+      end
 
-    it 'uses memoized errors' do
-      expect_any_instance_of(described_class).to receive(:errors).once.and_call_original
-      2.times { error_serializer }
+      it 'returns errors as serialized json with jsonapi specification' do
+        expect(JSON.parse(error_serializer)).to eq(json_schema)
```
Comment (bestwebua) [inline spec/concepts/api/v1/serializer/error_spec.rb]: Use json matchers instead.

File: spec/concepts/api/v1/serializer/error_spec.rb [hunk 7]
```diff
@@ -8,23 +8,34 @@
...
-              'detail' => detail
-            }
-          ]
-        }
-      )
+    context 'when error is a Hash' do
+      let(:object) { { errors: [{ pointer => detail }] } }
+
+      it 'returns errors as serialized json with jsonapi specification' do
+        expect(JSON.parse(error_serializer)).to eq(json_schema)
```
Comment (bestwebua) [inline spec/concepts/api/v1/serializer/error_spec.rb]: The same.

File: spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    let(:user) { create(:user) }
+    let(:user_data) { { id: user.id, email: user.email, created_at: user.created_at } }
+    let(:email_token) { Api::V1::Lib::Service::EmailToken.encode(user_data) }
+    let(:params) { { email_token: email_token } }
+
+    describe 'Success' do
+      it 'returns updated user' do
+        expect(operation.success).to eq(true)
+        expect(operation).to be_success
+        expect(User.find(user.id)).to be_confirmed
```
Comment (bestwebua) [inline spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb]: expect(user.reload).to be_confirmed

File: spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    describe 'Failure' do
+      context 'when user with invalid token' do
+        let(:email_token) { 'this.is.token' }
+
+        it 'returns errors' do
+          expect(operation.failure).to be_an_instance_of(::Hash)
+          expect(operation.failure[:errors]).to include({ token: I18n.t('users.confirmations.token.invalid') })
+          expect(User.find(user.id)).not_to be_confirmed
```
Comment (bestwebua) [inline spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb]: expect(user.reload).not_to be_confirmed

File: spec/concepts/api/v1/users/registrations/operation/create_spec.rb [hunk 10]
```diff
@@ -10,7 +10,7 @@
 
     describe 'Success' do
       it 'returns saved user' do
-        expect { operation }.to change(User, :count).from(0).to(1)
+        expect { operation }.to change(User, :count)
```
Comment (bestwebua) [inline spec/concepts/api/v1/users/registrations/operation/create_spec.rb]: .from(0).to(1) is more explicity

File: app/concepts/api/v1/serializer/error.rb [hunk 11]
```diff
@@ -15,11 +15,20 @@ def to_json
     attr_reader :object
 
     def errors
-      @errors ||= @object.errors.map do |error|
-        {
-          source: { pointer: "/data/attributes/#{error.path.first}" },
-          detail: error.text
-        }
+      if @object.is_a?(::Hash)
```
Comment (bestwebua) [inline app/concepts/api/v1/serializer/error.rb:18]: Guard clauses?

File: app/concepts/api/v1/users/confirmations/operation/confirm.rb (hunk 5 above)
Comment (bestwebua) [reply app/concepts/api/v1/users/confirmations/operation/confirm.rb]: Guard clauses?

File: app/controllers/concerns/response.rb [hunk 12]
```diff
@@ -1,7 +1,11 @@
 # frozen_string_literal: true
 
 module Response
-  def respond_with(entity:, status: 200, serializer: Api::V1::Serializer::Error)
-    render json: serializer.new(entity).to_json, status: status
+  def respond_with(entity: nil, status: 200, serializer: Api::V1::Serializer::Error)
+    if entity
```
Comment (bestwebua) [inline app/controllers/concerns/response.rb]: Guard clauses?

File: spec/concepts/api/v1/serializer/error_spec.rb [hunk 13]
```diff
@@ -8,23 +8,34 @@
     let(:error_object) { instance_double('ErrorleObject', path: [pointer], text: detail) }
     let(:pointer) { :some_pointer }
     let(:detail) { 'some detail' }
+    let(:json_schema) do
```
Comment (bestwebua) [inline spec/concepts/api/v1/serializer/error_spec.rb]: Honestly, it's not a json schema.

File: spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb [hunk 14]
```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+RSpec.describe Api::V1::Users::Confirmations::Operation::Confirm do
+  describe '.call' do
+    subject(:operation) { described_class.call(params) }
+
+    let(:user) { create(:user) }
+    let(:user_data) { { id: user.id, email: user.email, created_at: user.created_at } }
+    let(:email_token) { Api::V1::Lib::Service::EmailToken.encode(user_data) }
```
Comment (bestwebua) [inline spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb]: rspec helpers???

File: spec/requests/v1/users/confirmations_spec.rb [hunk 15]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+RSpec.describe 'Confirmations', type: :request do
+  let(:user) { create(:user) }
+  let(:user_data) { { id: user.id, email: user.email, created_at: user.created_at } }
+  let(:email_token) { Api::V1::Lib::Service::EmailToken.encode(user_data) }
```
Comment (bestwebua) [inline spec/requests/v1/users/confirmations_spec.rb]: rspec helper?

File: app/controllers/concerns/response.rb (hunk 12 above)
Comment (DandyCV) [reply app/controllers/concerns/response.rb]: Sure!

File: app/concepts/api/v1/serializer/error.rb (hunk 11 above)
Comment (DandyCV) [reply app/concepts/api/v1/serializer/error.rb:18]: Don't think it's necessary here.

File: spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb (hunk 8 above)
Comment (DandyCV) [reply spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb]: Thanks, I'll use it.

File: spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb (hunk 14 above)
Comment (DandyCV) [reply spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb]: Done

File: spec/requests/v1/users/confirmations_spec.rb (hunk 15 above)
Comment (DandyCV) [reply spec/requests/v1/users/confirmations_spec.rb]: Done

File: spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb (hunk 9 above)
Comment (DandyCV) [reply spec/concepts/api/v1/users/confirmations/operation/confirm_spec.rb]: 👍

File: spec/concepts/api/v1/users/registrations/operation/create_spec.rb (hunk 10 above)
Comment (DandyCV) [reply spec/concepts/api/v1/users/registrations/operation/create_spec.rb]: 👍

File: app/concepts/api/v1/users/confirmations/operation/confirm.rb (hunk 4 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/confirmations/operation/confirm.rb]: Yes, my bad!

File: spec/concepts/api/v1/serializer/error_spec.rb (hunk 13 above)
Comment (DandyCV) [reply spec/concepts/api/v1/serializer/error_spec.rb]: I agree.I have replaced this with error_hash

File: spec/concepts/api/v1/serializer/error_spec.rb (hunk 6 above)
Comment (DandyCV) [reply spec/concepts/api/v1/serializer/error_spec.rb]: This is not json, replaced with error_hash

File: spec/concepts/api/v1/serializer/error_spec.rb (hunk 7 above)
Comment (DandyCV) [reply spec/concepts/api/v1/serializer/error_spec.rb]: ^^

File: spec/concepts/api/v1/users/confirmations/operation/show_spec.rb [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+    let(:user) { create(:user) }
+    let(:user_data) { { id: user.id, email: user.email, created_at: user.created_at } }
+    let(:email_token) { generate_token(user_data) }
+    let(:params) { { email_token: email_token } }
+
+    before { stub_const('Api::V1::Lib::Service::EmailToken::HMAC_SECRET', 'test123456') }
+
+    describe 'Success' do
+      it 'returns updated user' do
```
Comment (bestwebua) [inline spec/concepts/api/v1/users/confirmations/operation/show_spec.rb:15]: confirmates target user...

expect { operation }.to change { user.reload.confirmed }.from(false).to(true)....

**Review by bestwebua — APPROVED** [review]

**Pull request merged**